encrypted-config-value/src/main/java/com/palantir/config/crypto/value/AesEncryptedValue.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.config.crypto.value;
+
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.util.Arrays;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@JsonDeserialize(as = ImmutableAesEncryptedValue.class)
+@JsonSerialize(as = ImmutableAesEncryptedValue.class)
+@JsonTypeName("AES")
+public abstract class AesEncryptedValue extends EncryptedValue {
+    private static final int LEGACY_IV_SIZE = 256 / Byte.SIZE;
+    private static final int LEGACY_TAG_SIZE = 128 / Byte.SIZE;
+
+    public static AesEncryptedValue fromLegacy(LegacyEncryptedValue value) {
+        byte[] buf = value.getCiphertext();
+        byte[] iv = Arrays.copyOfRange(buf, 0, LEGACY_IV_SIZE);
+        byte[] ct = Arrays.copyOfRange(buf, LEGACY_IV_SIZE, buf.length - LEGACY_TAG_SIZE);
+        byte[] tag = Arrays.copyOfRange(buf, buf.length - LEGACY_TAG_SIZE, buf.length);
+
+        return ImmutableAesEncryptedValue.builder()
+                .mode(Mode.GCM)
+                .iv(iv)
+                .ciphertext(ct)
+                .tag(tag)
+                .build();
+    }
+
+    public abstract Mode getMode();
+
+    public abstract byte[] getIv();
```

Comment:
One more question -- what controls the JSON serialization format of this object right now? Since byte arrays don't have a direct representation in JSON, assume that we need to define a serializer that turns the byte array fields into base64-encoded strings. That would mean that raw JSON would be:

```
{
   "mode": ...,
   "iv:"<b64-encoded-iv>",
   "ciphertext":"<b64-encoded-ciphertext>",
...
}
```

Then the "serialized" representation is the B64-encoded version of this JSON string.

Apologies if you're already doing this somewhere, but looked around and didn't see any code that seems to do this in the PR (I see that there's a call to `return MAPPER.writeValueAsBytes(aesEncryptedValue);`, but can't tell how this call will handle the `byte[]` fields).